cli/tools/bench.rs

```diff
@@ -0,0 +1,727 @@
+      summary.ignored,
+      summary.measured,
+      summary.filtered_out,
+      colors::gray(format!("({})", display::human_elapsed(elapsed.as_millis()))),
+    );
+  }
+}
+
+fn create_reporter(echo_output: bool) -> Box<dyn BenchReporter + Send> {
+  Box::new(PrettyBenchReporter::new(echo_output))
+}
+
+/// Type check a collection of module and document specifiers.
+async fn check_specifiers(
+  ps: &ProcState,
+  permissions: Permissions,
+  specifiers: Vec<ModuleSpecifier>,
+  lib: emit::TypeLib,
+) -> Result<(), AnyError> {
+  ps.prepare_module_load(
+    specifiers,
+    false,
+    lib,
+    Permissions::allow_all(),
+    permissions,
+    true,
+  )
+  .await?;
+
+  Ok(())
+}
+
+/// Run a single specifier as an executable bench module.
+async fn bench_specifier(
+  ps: ProcState,
+  permissions: Permissions,
+  specifier: ModuleSpecifier,
+  channel: UnboundedSender<BenchEvent>,
+  options: BenchSpecifierOptions,
+) -> Result<(), AnyError> {
+  let mut worker = create_main_worker(
+    &ps,
+    specifier.clone(),
+    permissions,
+    vec![ops::bench::init(channel.clone())],
+  );
+
+  if options.compat_mode {
+    worker.execute_side_module(&compat::GLOBAL_URL).await?;
+    worker.execute_side_module(&compat::MODULE_URL).await?;
+
+    let use_esm_loader = compat::check_if_should_use_esm_loader(&specifier)?;
+
+    if use_esm_loader {
+      worker.execute_side_module(&specifier).await?;
+    } else {
+      compat::load_cjs_module(
+        &mut worker.js_runtime,
+        &specifier.to_file_path().unwrap().display().to_string(),
+        false,
+      )?;
+      worker.run_event_loop(false).await?;
+    }
+  } else {
+    // We execute the module module as a side module so that import.meta.main is not set.
+    worker.execute_side_module(&specifier).await?;
+  }
+
+  worker.dispatch_load_event(&located_script_name!())?;
+
+  let bench_result = worker.js_runtime.execute_script(
+    &located_script_name!(),
+    &format!(
+      r#"Deno[Deno.internal].runBenchmarks({})"#,
+      json!({
+        "filter": options.filter,
+      }),
+    ),
+  )?;
+
+  worker.js_runtime.resolve_value(bench_result).await?;
+
+  worker.dispatch_unload_event(&located_script_name!())?;
+
+  Ok(())
+}
+
+/// Test a collection of specifiers with test modes concurrently.
+async fn bench_specifiers(
+  ps: ProcState,
+  permissions: Permissions,
+  specifiers: Vec<ModuleSpecifier>,
+  options: BenchSpecifierOptions,
+) -> Result<(), AnyError> {
+  let log_level = ps.flags.log_level;
+
+  let (sender, mut receiver) = unbounded_channel::<BenchEvent>();
+
+  let join_handles = specifiers.iter().map(move |specifier| {
+    let ps = ps.clone();
+    let permissions = permissions.clone();
+    let specifier = specifier.clone();
+    let sender = sender.clone();
+    let options = options.clone();
+
+    tokio::task::spawn_blocking(move || {
+      let future = bench_specifier(ps, permissions, specifier, sender, options);
+
+      run_basic(future)
```

Comment:
I was just looking at the test code yesterday with Kitson and I don't understand why we create a new tokio runtime for each specifier. Can't we just do `spawn(async move { bench_specifier(...).await })` here?

Reply:
Because we have `--jobs` flag that allows to run testing code on multiple threads. In such case sync code would block progress in different specifiers.

Reply:
I guess we could do that for `bench` though; trying to think if there's a case that would prohibit using `tokio::task::spawn`

Reply:
Hmm... don't think we need to create a completely new runtime for that. If we're blocking then we could transition the task to be blocking?

Reply:
Nevermind, just had a call with Bartek and he explained to me the concept of tokio's single threaded runtime and how we're using that so we can't don't this.